In [61]:
from keras.layers import Conv1D, Dense, BatchNormalization, ReLU, GlobalAveragePooling1D, LeakyReLU, Input, LayerNormalization, Dropout
from keras import Sequential
from keras.callbacks import ModelCheckpoint, ReduceLROnPlateau, EarlyStopping
from stock_data_ingestion import ingest_stocks_to_df
from preprocessing_ts_keras_finance import to_ts_data
from keras.utils import plot_model

In [42]:
## comment first line out after first use to prevent 
## giving the source website for the stocks too much traffic. 
stocks_df = ingest_stocks_to_df()
stock_list = stocks_df.company_name.unique()
# ## displays
# display(stocks_df)
# for stock in stock_list:
#     print(stock, 'len', len(stocks_df[stocks_df.company_name == stock].index))
# sns.lineplot(x=stocks_df.index, y='Adj Close',data=stocks_df, legend='auto', hue='company_name')
# plt.show()
data_gen_train, data_gen_test = to_ts_data(stock_list[0], stocks_df) 

In [74]:
def conv_ts(dropout=0.5):
    model = Sequential()
    model.add(Conv1D(filters=64, kernel_size=3, padding="same"))
    model.add(LayerNormalization())
    model.add(ReLU())
    model.add(Conv1D(filters=64, kernel_size=3, padding="same"))
    model.add(LayerNormalization())
    model.add(Dropout(dropout))
    model.add(ReLU())
    model.add(Conv1D(filters=64, kernel_size=3, padding="same"))
    model.add(LayerNormalization())
    model.add(ReLU())
    model.add(GlobalAveragePooling1D())
    model.add(Dense(1))
    model.add(LeakyReLU(alpha=1))
    model.compile(optimizer="adam", loss="mse", metrics=["mse"],)
    return model

In [76]:
callbacks = [
    ModelCheckpoint(
        "best_model.h5", save_best_only=True, monitor="loss"
    ),
    ReduceLROnPlateau(
        monitor="loss", factor=0.5, patience=20, min_lr=0.0001
    ),
    EarlyStopping(monitor="loss", patience=50, verbose=1),
]

In [77]:
epochs = 700
model = conv_ts()
model.fit(data_gen_train, epochs=epochs,
    callbacks=callbacks,
    verbose=1)


Epoch 1/700
9/9 [==============================] - 2s 87ms/step - loss: 17.3596 - mse: 17.3596
Epoch 2/700
9/9 [==============================] - 1s 93ms/step - loss: 7.7699 - mse: 7.7699
Epoch 3/700
9/9 [==============================] - 1s 90ms/step - loss: 3.9739 - mse: 3.9739
Epoch 4/700
9/9 [==============================] - 1s 84ms/step - loss: 2.5457 - mse: 2.5457
Epoch 5/700
9/9 [==============================] - 1s 88ms/step - loss: 3.2533 - mse: 3.2533
Epoch 6/700
9/9 [==============================] - 1s 85ms/step - loss: 2.3838 - mse: 2.3838
Epoch 7/700
9/9 [==============================] - 1s 84ms/step - loss: 3.0384 - mse: 3.0384
Epoch 8/700
9/9 [==============================] - 1s 86ms/step - loss: 3.1688 - mse: 3.1688
Epoch 9/700
9/9 [==============================] - 1s 84ms/step - loss: 2.5503 - mse: 2.5503
Epoch 10/700
9/9 [==============================] - 1s 88ms/step - loss: 2.7639 - mse: 2.7639
Epoch 11/700
9/9 [==============================] - 1s 90ms/step - 

9/9 [==============================] - 1s 111ms/step - loss: 2.5784 - mse: 2.5784
Epoch 89/700
9/9 [==============================] - 1s 135ms/step - loss: 2.2445 - mse: 2.2445
Epoch 90/700
9/9 [==============================] - 1s 129ms/step - loss: 2.9662 - mse: 2.9662
Epoch 91/700
9/9 [==============================] - 1s 89ms/step - loss: 3.2313 - mse: 3.2313
Epoch 92/700
9/9 [==============================] - 1s 77ms/step - loss: 3.1612 - mse: 3.1612
Epoch 93/700
9/9 [==============================] - 1s 76ms/step - loss: 2.9425 - mse: 2.9425
Epoch 94/700
9/9 [==============================] - 1s 85ms/step - loss: 2.5592 - mse: 2.5592
Epoch 95/700
9/9 [==============================] - 1s 89ms/step - loss: 3.2981 - mse: 3.2981
Epoch 96/700
9/9 [==============================] - 1s 75ms/step - loss: 3.2060 - mse: 3.2060
Epoch 97/700
9/9 [==============================] - 1s 73ms/step - loss: 2.1769 - mse: 2.1769
Epoch 98/700
9/9 [==============================] - 1s 74ms/step - los

9/9 [==============================] - 1s 133ms/step - loss: 2.8765 - mse: 2.8765
Epoch 175/700
9/9 [==============================] - 1s 95ms/step - loss: 3.1863 - mse: 3.1863
Epoch 176/700
9/9 [==============================] - 1s 104ms/step - loss: 3.0810 - mse: 3.0810
Epoch 177/700
9/9 [==============================] - 1s 103ms/step - loss: 2.9893 - mse: 2.9893
Epoch 178/700
9/9 [==============================] - 2s 181ms/step - loss: 2.8651 - mse: 2.8651
Epoch 179/700
9/9 [==============================] - 2s 197ms/step - loss: 2.5190 - mse: 2.5190
Epoch 180/700
9/9 [==============================] - 2s 208ms/step - loss: 2.0040 - mse: 2.0040
Epoch 181/700
9/9 [==============================] - 2s 194ms/step - loss: 3.1597 - mse: 3.1597
Epoch 182/700
9/9 [==============================] - 1s 117ms/step - loss: 2.3814 - mse: 2.3814
Epoch 183/700
9/9 [==============================] - 1s 87ms/step - loss: 1.9843 - mse: 1.9843
Epoch 184/700
9/9 [==============================] - 1s 

9/9 [==============================] - 1s 77ms/step - loss: 1.9182 - mse: 1.9182
Epoch 260/700
9/9 [==============================] - 1s 73ms/step - loss: 2.6387 - mse: 2.6387
Epoch 261/700
9/9 [==============================] - 1s 72ms/step - loss: 2.1095 - mse: 2.1095
Epoch 262/700
9/9 [==============================] - 1s 73ms/step - loss: 2.2434 - mse: 2.2434
Epoch 263/700
9/9 [==============================] - 1s 78ms/step - loss: 2.2357 - mse: 2.2357
Epoch 264/700
9/9 [==============================] - 1s 96ms/step - loss: 2.3433 - mse: 2.3433
Epoch 265/700
9/9 [==============================] - 1s 108ms/step - loss: 2.0725 - mse: 2.0725
Epoch 266/700
9/9 [==============================] - 1s 93ms/step - loss: 1.8388 - mse: 1.8388
Epoch 267/700
9/9 [==============================] - 1s 82ms/step - loss: 2.0670 - mse: 2.0670
Epoch 268/700
9/9 [==============================] - 1s 79ms/step - loss: 2.0068 - mse: 2.0068
Epoch 269/700
9/9 [==============================] - 1s 77ms/st

9/9 [==============================] - 1s 77ms/step - loss: 2.0132 - mse: 2.0132
Epoch 346/700
9/9 [==============================] - 1s 71ms/step - loss: 2.2577 - mse: 2.2577
Epoch 347/700
9/9 [==============================] - 1s 72ms/step - loss: 1.5806 - mse: 1.5806
Epoch 348/700
9/9 [==============================] - 1s 75ms/step - loss: 1.7324 - mse: 1.7324
Epoch 349/700
9/9 [==============================] - 1s 72ms/step - loss: 1.3765 - mse: 1.3765
Epoch 350/700
9/9 [==============================] - 1s 72ms/step - loss: 2.4699 - mse: 2.4699
Epoch 351/700
9/9 [==============================] - 1s 72ms/step - loss: 1.6890 - mse: 1.6890
Epoch 352/700
9/9 [==============================] - 1s 74ms/step - loss: 2.9363 - mse: 2.9363
Epoch 353/700
9/9 [==============================] - 1s 71ms/step - loss: 1.6458 - mse: 1.6458
Epoch 354/700
9/9 [==============================] - 1s 83ms/step - loss: 2.3325 - mse: 2.3325
Epoch 355/700
9/9 [==============================] - 1s 108ms/st

9/9 [==============================] - 1s 74ms/step - loss: 1.5448 - mse: 1.5448
Epoch 432/700
9/9 [==============================] - 1s 76ms/step - loss: 1.9303 - mse: 1.9303
Epoch 433/700
9/9 [==============================] - 1s 76ms/step - loss: 1.9305 - mse: 1.9305
Epoch 434/700
9/9 [==============================] - 1s 74ms/step - loss: 1.8002 - mse: 1.8002
Epoch 435/700
9/9 [==============================] - 1s 75ms/step - loss: 1.9014 - mse: 1.9014
Epoch 436/700
9/9 [==============================] - 1s 74ms/step - loss: 2.0474 - mse: 2.0474
Epoch 437/700
9/9 [==============================] - 1s 77ms/step - loss: 1.8188 - mse: 1.8188
Epoch 438/700
9/9 [==============================] - 1s 73ms/step - loss: 1.3428 - mse: 1.3428
Epoch 439/700
9/9 [==============================] - 1s 73ms/step - loss: 1.3599 - mse: 1.3599
Epoch 440/700
9/9 [==============================] - 1s 79ms/step - loss: 1.7090 - mse: 1.7090
Epoch 441/700
9/9 [==============================] - 1s 88ms/ste

9/9 [==============================] - 1s 85ms/step - loss: 1.3385 - mse: 1.3385
Epoch 518/700
9/9 [==============================] - 1s 92ms/step - loss: 1.4951 - mse: 1.4951
Epoch 519/700
9/9 [==============================] - 1s 87ms/step - loss: 1.5187 - mse: 1.5187
Epoch 520/700
9/9 [==============================] - 1s 96ms/step - loss: 1.3026 - mse: 1.3026
Epoch 521/700
9/9 [==============================] - 1s 90ms/step - loss: 1.9187 - mse: 1.9187
Epoch 522/700
9/9 [==============================] - 1s 104ms/step - loss: 1.3275 - mse: 1.3275
Epoch 523/700
9/9 [==============================] - 1s 92ms/step - loss: 1.2989 - mse: 1.2989
Epoch 524/700
9/9 [==============================] - 1s 110ms/step - loss: 1.2617 - mse: 1.2617
Epoch 525/700
9/9 [==============================] - 1s 93ms/step - loss: 0.9070 - mse: 0.9070
Epoch 526/700
9/9 [==============================] - 1s 78ms/step - loss: 1.6712 - mse: 1.6712
Epoch 527/700
9/9 [==============================] - 1s 78ms/s

Epoch 603/700
9/9 [==============================] - 1s 121ms/step - loss: 2.7146 - mse: 2.7146
Epoch 604/700
9/9 [==============================] - 1s 128ms/step - loss: 1.3920 - mse: 1.3920
Epoch 605/700
9/9 [==============================] - 1s 124ms/step - loss: 1.1449 - mse: 1.1449
Epoch 606/700
9/9 [==============================] - 1s 141ms/step - loss: 0.8906 - mse: 0.8906
Epoch 607/700
9/9 [==============================] - 1s 129ms/step - loss: 1.0746 - mse: 1.0746
Epoch 608/700
9/9 [==============================] - 1s 114ms/step - loss: 1.0884 - mse: 1.0884
Epoch 609/700
9/9 [==============================] - 1s 120ms/step - loss: 0.6642 - mse: 0.6642
Epoch 610/700
9/9 [==============================] - 1s 138ms/step - loss: 0.7482 - mse: 0.7482
Epoch 611/700
9/9 [==============================] - 1s 115ms/step - loss: 0.8511 - mse: 0.8511
Epoch 612/700
9/9 [==============================] - 1s 135ms/step - loss: 1.1848 - mse: 1.1848
Epoch 613/700
9/9 [=====================

9/9 [==============================] - 1s 92ms/step - loss: 0.6322 - mse: 0.6322
Epoch 689/700
9/9 [==============================] - 1s 94ms/step - loss: 0.7022 - mse: 0.7022
Epoch 690/700
9/9 [==============================] - 1s 98ms/step - loss: 0.6479 - mse: 0.6479
Epoch 691/700
9/9 [==============================] - 1s 154ms/step - loss: 0.9061 - mse: 0.9061
Epoch 692/700
9/9 [==============================] - 1s 145ms/step - loss: 1.0090 - mse: 1.0090
Epoch 693/700
9/9 [==============================] - 1s 131ms/step - loss: 1.3426 - mse: 1.3426
Epoch 694/700
9/9 [==============================] - 1s 103ms/step - loss: 2.0011 - mse: 2.0011
Epoch 695/700
9/9 [==============================] - 1s 98ms/step - loss: 1.3151 - mse: 1.3151
Epoch 696/700
9/9 [==============================] - 1s 94ms/step - loss: 1.1091 - mse: 1.1091
Epoch 697/700
9/9 [==============================] - 1s 92ms/step - loss: 0.8587 - mse: 0.8587
Epoch 698/700
9/9 [==============================] - 1s 98ms

In [78]:
model.evaluate(data_gen_test)

2/2 [==============================] - 0s 27ms/step - loss: 0.5011 - mse: 0.5011


[0.5010643005371094, 0.5010643005371094]

- with layer, no dropout: test: mse: 0.6667548418045044, 500 epochs
- with layer, 0.5 dropout second layer: [0.3815690279006958, 0.3815690279006958], ~1.97 train 500 epochs; second run: 0.5011, 0.6579
